### PR TITLE
Fix #260 implement addPost and getPost method

### LINKED ISSUE #260: Allow a Post to get stored in Redis
In #201 we added [storage.js](src/storage.js) and some initial [tests](test/storage.test.js).  We need to extend that to allow for a Post to be saved and returned from Redis.  We can (eventually) connect this with the work happening in #259 to add a `Post` type, accepting and returning `Post` objects.

We need `addPost()` and `getPost()` methods to be added to the `storage` module.  The code can be similar to what we have now for `addFeed()` and `getFeed()`.  In Redis, a `Post` can be a Hash, with all the different properties in the `Post` object corresponding to fields in the Hash--like the feed does with `name` and `url`.

Each `Post` in Redis will need a unique key.  Let's start by using a `Post`'s `guid` property, and see if that works well enough.

We'll also want to create a `posts` Sorted Set in Redis to hold all the `guid` keys for the posts that we store.  Adding an item to the `posts` set should include the `published` time as a score. In other words, we'll sort the posts in our set according to the date they were written, so it's easy to get them out in the order we want.

See https://redis.io/topics/data-types and other docs on how the Hash and Sorted Set data types work in Redis, and https://github.com/luin/ioredis for docs on how to call them from JavaScript.

### PR DESCRIPTION
**Pull request by KarlaChen**
Added addPost() function using guid as unique key for each post and store posts as sorted set by published date.
Added getPost() by guid and getPosts() by range of dates.
Added test for testing above mentioned functions.

### CODE REVIEW

File: src/storage.js [hunk 1]
```diff
@@ -23,4 +26,41 @@ module.exports = {
...
 
   getFeed: feedID => redis.hgetall(feedID),
+
+  //  addPost
+  addPost: async post => {
+    const postId = await redis.incr(POST_ID);
+    await redis
+      .multi()
+      .hmset(
+        postId,
```
Comment (humphd) [inline src/storage.js]: Let's try using `post.guid` here, and not bother with `postId` or `POST_ID` above.  I think the `guid` should be unique enough to use here, but we can experiment.

File: src/storage.js [hunk 2]
```diff
@@ -4,8 +4,11 @@ const { redis } = require('./lib/redis');
 const FEED_ID = 'feed_id';
 const FEEDS = 'feeds';
 
+const POST_ID = 'post_id';
+const POSTS = 'posts';
+
 module.exports = {
-  addFeed: async (name, url) => {
+  addFeed: async (author, url) => {
```
Comment (humphd) [inline src/storage.js]: Why this change?  It's not wrong, but it seems unrelated to this fix.

File: src/storage.js [hunk 3]
```diff
@@ -4,8 +4,11 @@ const { redis } = require('./lib/redis');
 const FEED_ID = 'feed_id';
 const FEEDS = 'feeds';
 
+const POST_ID = 'post_id';
```
Comment (humphd) [inline src/storage.js]: I bet we can get rid of this, see my comment below.

File: src/storage.js [hunk 4]
```diff
@@ -23,4 +26,41 @@ module.exports = {
   getFeeds: () => redis.smembers(FEEDS),
 
   getFeed: feedID => redis.hgetall(feedID),
+
+  //  addPost
+  addPost: async post => {
+    const postId = await redis.incr(POST_ID);
```
Comment (humphd) [inline src/storage.js]: Let's drop this.

File: src/storage.js [hunk 5]
```diff
@@ -23,4 +26,41 @@ module.exports = {
   getFeeds: () => redis.smembers(FEEDS),
 
   getFeed: feedID => redis.hgetall(feedID),
+
+  //  addPost
```
Comment (humphd) [inline src/storage.js]: This comment isn't adding anything, let's remove it.

File: src/storage.js [hunk 6]
```diff
@@ -23,4 +26,41 @@ module.exports = {
...
+  addPost: async post => {
+    const postId = await redis.incr(POST_ID);
+    await redis
+      .multi()
+      .hmset(
+        postId,
+        'guid',
```
Comment (humphd) [inline src/storage.js]: If we store the post by `guid`, we probably don't need to store it in the hash, this can be removed.

File: src/storage.js [hunk 7]
```diff
@@ -23,4 +26,41 @@ module.exports = {
...
+        'updated',
+        post.updated,
+        'published',
+        post.published,
+        'url',
+        post.url,
+        'site',
+        post.site
+      )
+      .zadd(POSTS, Date.parse(post.published), postId)
```
Comment (humphd) [inline src/storage.js]: I think that we can assume that `post.published` will already be a date vs. a date string.  Also, you can use `post.guid` instead of `postId`.

File: src/storage.js [hunk 8]
```diff
@@ -23,4 +26,41 @@ module.exports = {
...
+      .exec();
+    return postId;
```
Comment (humphd) [inline src/storage.js]: I don't think we need to return anything, since we'll already know the key for this post, namely, the `post.guid`

File: src/storage.js [hunk 9]
```diff
@@ -23,4 +26,41 @@ module.exports = {
...
+  },
+
+  getPosts: () => redis.zrangebyscore(POSTS, 0, +Infinity),
```
Comment (humphd) [inline src/storage.js]: This is great.  Either in this PR, or in a follow-up, we can support paging here, by allowing the caller to pass arguments to specify which elements in the set to get.  That way they can request a portion vs. everything.

File: src/storage.js (hunk 2 above)
Comment (haichuan0424) [reply src/storage.js]: @humphd When we doing the test, we found weird error that the name property got passed into the post object. After we changed this name to match the property name inside the post object, the error disappeared.

Comment (humphd) [reply src/storage.js]: If that's true, we should probably file a bug with ioredis or something.  For now, leave it as `author`, but please show me when I see you next time.

File: src/storage.js (hunk 7 above)
Comment (haichuan0424) [reply src/storage.js]: @humphd In order to sort the set, the score value must be numeric. The Date.parse function will transfer the date string to an integer millisecond value based on how far ahead to a base time.

Comment (humphd) [reply src/storage.js]: Agree.  I'm saying, assume you have a `Date` and just call `getTime()` to get the `ms`:

```
> new Date()
2019-11-23T02:53:01.273Z
> new Date().getTime()
1574477586738
```

File: src/storage.js (hunk 2 above)
Comment (haichuan0424) [reply src/storage.js]: Sure, will show you next week.

File: src/storage.js (hunk 7 above)
Comment (haichuan0424) [reply src/storage.js]: OK, we can try this.

File: src/storage.js (hunk 2 above)
Comment (KarlaChen) [reply src/storage.js]: Here is the error if we keep using "name"
image

Comment (humphd) [reply src/storage.js]: I pulled your branch, and figured out the issue.  It's not `name` vs. `author`.  The problem is that you are re-using keys in Redis.  As I mentioned in my review, we need to use the `post.guid` as the key for a Post Hash, and to put into the sorted set.  You are using two auto-incrementing values, and the feedId and postId are colliding, causing the data to be missing in your tests.

If you implement the changes I requested, this should be fixed.

**Comment by humphd** [comment]
That failure is #343, not related to your changes.

File: src/storage.js [hunk 10]
```diff
@@ -23,4 +25,39 @@ module.exports = {
...
+        'updated',
+        post.updated,
+        'published',
+        post.published,
+        'url',
+        post.url,
+        'site',
+        post.site
+      )
+      .zadd(POSTS, new Date(post.published).getTime(), post.guid) // sort set by published date as scores
```
Comment (humphd) [inline src/storage.js]: I'm still worried about `post.published` here.  I think it should already be a `Date`, and you should just call `getTime()` on it.  cc @jatinAroraGit, who is working on the `Post` class.

File: src/storage.js [hunk 11]
```diff
@@ -23,4 +25,39 @@ module.exports = {
...
+      .exec();
+  },
+
+  getPosts: (startDate, endDate) =>
```
Comment (mskuybeda) [inline src/storage.js]: Not sure about startDate and endDate. You could just pass it as Date parameter.

Comment (manekenpix) [reply src/storage.js]: I agree with @mskuybeda, I think we could pass a Date and get everything between that date and [the current date](https://javascript.info/date#creation).
